nntrainer/compiler/tflite_interpreter.cpp

```diff
@@ -199,30 +200,114 @@ class TfOpNode {
-class TfOpIdxMap {
+template <typename T> class BidirectionalIndexMap {
 public:
-  TfOpIdxMap(std::vector<TfOpNode> nodes){};
+  /**
+   * @brief addDatapoint to the map
+   *
+   * @param data data to be added if there is no occurrence, data will be
+   * copied.
+   */
+  void addDataWhenNotFound(T data) {
+    auto search = data2index.find(data);
+
+    if (search == data2index.end()) {
+      data2index[data] = index2data.size();
+      index2data.push_back(data);
+    }
+  }
 
-private:
   /**
-   * @brief Bidirectional Index map
+   * @brief Get the Index of the data
    *
-   * @tparam T type of a underyling value
+   * @param key data that will be the key
+   * @return unsigned int index
    */
-  template <typename T> class BidirectionalIndexMap {
-    std::unordered_map<T, unsigned int> data2index; /**< data -> index map */
-    std::vector<T> index2data;                      /**< index -> data map */
-  };
+  unsigned int getIndex(const T &key) const {
+    auto search = data2index.find(key);
+
+    NNTR_THROW_IF(search == data2index.end(), std::invalid_argument)
+      << FUNC_TAG << "Cannot find index for key: " << key;
 
+    return *search;
+  }
+
+  /**
+   * @brief Get the Data object
+   *
+   * @param idx index to be searched
+   * @return T datapoint T
+   */
+  T getData(unsigned int index) const {
+    NNTR_THROW_IF(index >= index2data.size(), std::invalid_argument)
+      << FUNC_TAG << "Cannot find data for index: " << index;
+
+    return index2data[index];
+  }
+
+private:
+  std::unordered_map<T, unsigned int> data2index; /**< data -> index map */
+  std::vector<T> index2data;                      /**< index -> data map */
+};
+
+/**
+ * @brief tensorflow operation index map, this class manages operation index
+ * mapping
+ *
+ */
+class TfOpIdxMap {
+public:
+  TfOpIdxMap(const TfOpNodes &nodes) {
+    auto &opcode_map = getIndexMap<tflite::BuiltinOperator>();
+    auto update_opcode = [&opcode_map](tflite::BuiltinOperator opcode) {
+      opcode_map.addDataWhenNotFound(opcode);
+    };
+
+    auto &buffer_map = getIndexMap<const float *>();
+    buffer_map.addDataWhenNotFound(empty_buffer); /// put empty buffer to first
+
+    auto update_buffers = [&buffer_map](const TfOpNode::Variables &variables) {
+      for (auto &variable : variables) {
+        const Tensor &t = variable->getVariableRef();
+        if (!t.uninitialized() && t.isAllocated()) {
+          buffer_map.addDataWhenNotFound(t.getData());
+        }
+      }
+    };
+
+    auto &variable_map = getIndexMap<const Var_Grad *>();
+    auto update_variables =
+      [&variable_map](const TfOpNode::Variables &variables) {
+        for (auto &variable : variables) {
+          variable_map.addDataWhenNotFound(variable);
+        }
+      };
+
+    for (auto &op_node : nodes) {
+      update_opcode(op_node.getOpType());
+      update_variables(op_node.getInputs());
+      update_variables(op_node.getOutputs());
+      update_variables(op_node.getWeights());
```

Comment:
would it better to create a different map for weights? if you want to later use this map for saving the model to tflite, it saves the trouble of extracting the weights. Also, I am not sure at this point, but weights might need different treatment like datatypes etc compared to inputs/outputs in later time.

Reply:
Maybe buffer map can serve that purpose, but then full var_grad might be needed than just buffer.

Reply:
Actually, as far as I remember, this is merged in few pr's ahead.

Reply:
got it.